Stop Copying Every Http Request in Message Handler

original-brownbear opened the pull request:
* Copying the request is not necessary for requests that don't escape their content `BytesReference` in a way that makes it referenced after the request was responded to. We can simply release it once the response has been generated and a lot of `Unpooled` allocations that way for many if not most requests. For now, this PR makes it so search- and bulk requests are not copied to `Unpooled` buffers.
* Relates #32228
   * I think the issue that prevented that PR  from being merged was solved by #39634 that moved the bulk index marker search to ByteBuf bulk access so the composite buffer shouldn't require many additional bounds checks  (I'd argue the bounds checks we add, we save when copying the composite buffer)
* I couldn't necessarily reproduce much of a speedup (though no slowdown either) from this change, but I could reproduce a very measurable reduction in GC time with e.g. Rally's PMC (4g heap node and bulk requests of size 5k saw a reduction in young GC time by ~10% for me and a ~50% reduction in old gen GC time) 
* This also improves the real memory circuit breaker efficacy quite a bit I think since it strengthens the correlation between request size and actual memory used by the request (since we don't have a bunch of dangling, to-be-GCed requests hanging around).
* Also somewhat relates #44484

original-brownbear committed:
Stop Copying Every Http Request in Message Handler

* Copying the request is not necessary here. We can simply release it once the response has been generated and a lot of `Unpooled` allocations that way
* Relates #32228
   * I think the issue that preventet that PR  that PR from being merged was solved by #39634 that moved the bulk index marker search to ByteBuf bulk access so the composite buffer shouldn't require many additional bounds checks  (I'd argue the bounds checks we add, we save when copying the composite buffer)
* I couldn't neccessarily reproduce much of a speedup from this change, but I could reproduce a very measureable reduction in GC time with e.g. Rally's PMC (4g heap node and bulk requests of size 5k saw a reduction in young GC time by ~10% for me)

elasticmachine commented:
Pinging @elastic/es-distributed

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into never-copy-http-buffer

original-brownbear committed:
fix

original-brownbear committed:
maybe fix

tbrooks8 commented:
This PR relies on the fact that all REST actions will be done with the request content by the time they send a response. I'm not sure that this is a safe assumption.

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into never-copy-http-buffer

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into never-copy-http-buffer

original-brownbear committed:
stop retaining reference to actual request in put pipeline rest action

original-brownbear commented:
@tbrooks8 appears there's only a single spot where this doesn't hold: https://github.com/elastic/elasticsearch/pull/44564/files#diff-05216fd3df413bda2df9486bce7a4e29R51

I would also argue that holding on to these bytes without copying is something that shouldn't be happening  implicitly anywhere. Otherwise the whole memory behavior of the REST layer becomes pretty unpredictable. In the case of storing the pipeline source it's ok to use the request body for that as a special case, but if we generally allow request contents to be referenced beyond responding to the request the whole notion of the circuit breaker becomes wrong imo.

original-brownbear committed:
add assertion

original-brownbear committed:
Merge remote-tracking branch 'elastic' into never-copy-http-buffer

original-brownbear committed:
add back empty line

original-brownbear committed:
Fix NIO as well

original-brownbear committed:
fix nio as well

original-brownbear commented:
Jenkins run elasticsearch-ci/packaging-sample

original-brownbear commented on server/src/main/java/org/elasticsearch/rest/action/ingest/RestPutPipelineAction.java:
```diff
@@ -47,7 +48,8 @@ public String getName() {
     public RestChannelConsumer prepareRequest(RestRequest restRequest, NodeClient client) throws IOException {
         Tuple<XContentType, BytesReference> sourceTuple = restRequest.contentOrSourceParam();
-        PutPipelineRequest request = new PutPipelineRequest(restRequest.param("id"), sourceTuple.v2(), sourceTuple.v1());
+        PutPipelineRequest request = new PutPipelineRequest(restRequest.param("id"),
```
As mentioned in the comments, can't escape the bytes here (these are held on to by the ingest logic internally) since they'll be released eventually so we copy.

original-brownbear committed:
add comment on copying pipeline source

tbrooks8 commented:
This probably requires a team discuss.

original-brownbear commented:
Sure, added `team-discuss` :)

original-brownbear commented:
Putting this back into `WIP`, thanks @tbrooks8 for pointing out that `org.elasticsearch.transport.netty4.ByteBufBytesReference#toBytesRef` is not safe here.

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into never-copy-http-buffer

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into never-copy-http-buffer

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into never-copy-http-buffer

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into never-copy-http-buffer

original-brownbear committed:
only fast-path search+bulk req.

original-brownbear committed:
quick and dirty

original-brownbear committed:
too ambitious

original-brownbear committed:
Merge branch 'never-copy-http-buffer' of github.com:original-brownbear/elasticsearch into never-copy-http-buffer

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into never-copy-http-buffer

original-brownbear committed:
optimize some more

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into never-copy-http-buffer

original-brownbear committed:
faster

original-brownbear commented:
Jenkins test this

original-brownbear committed:
stop releasing unpooled buffers

original-brownbear commented on plugins/transport-nio/src/main/java/org/elasticsearch/http/nio/NioHttpRequest.java:
```diff
@@ -40,24 +42,35 @@
+    }
+
+    private NioHttpRequest(FullHttpRequest request, int sequence, AtomicBoolean released) {
+        this(request, new HttpHeadersMap(request.headers()), sequence, released, request.content() instanceof UnpooledHeapByteBuf == false,
```
Technically we could do more involved detection of whether or not the buffer here really is unpooled like was done in https://github.com/elastic/elasticsearch/pull/32228 which might be nice for low-heap nodes that will often see unpooled composite buffers. I just added this check to avoid wasting memory in obvious spots now, but IMO we could optimize with the deeper check for unpooled composite buffers in a follow up.

tbrooks8 replied:
I don't think this is reliable. In Nio these might often be `UnpooledHeapByteBuf` even if they are pooled from BigArrays. It's how we interface with big arrays (`PagedByteBuf`). I think in Nio for right now we should just always do the copy. As in remove the "pooled" optimization.

original-brownbear replied:
Makes sense, reverted the NIO implementation for now.

original-brownbear commented on server/src/main/java/org/elasticsearch/rest/action/document/RestBulkAction.java:
```diff
@@ -103,4 +103,9 @@ public RestChannelConsumer prepareRequest(final RestRequest request, final NodeC
+
+    @Override
+    public boolean allowsUnsafeBuffers() {
+        return true;
```
We could optimize this and other spots where we return `true` even further by releasing the request at the earliest possible point instead of when sending out the response. I'd push that to a follow up though, as it's not entirely trivial to get this right.

original-brownbear commented:
@ywelsch @tbrooks8 I've now made this logic only avoid the copying for bulk and search requests as discussed and added the necessary extension points to be more adaptive regarding when we release the request. I think this should be good for a review now :)

original-brownbear commented on modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpRequest.java:
```diff
@@ -105,9 +118,34 @@ public String uri() {
+        if (pooled && released.compareAndSet(false, true)) {
+            request.release();
+        }
+        content = null;
```
We don't allow getting `content` after release has been called now -> no point in keeping a reference to a potentially huge object here in case of unpooled buffers.

original-brownbear committed:
nicer assertion

original-brownbear commented:
Jenkins run elasticsearch-ci/1
Jenkins run elasticsearch-ci/2

(failed to download geoip file)

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into never-copy-http-buffer

original-brownbear commented:
@ywelsch @tbrooks8 ping, just in case you find yourself with a sec to take a look :) Thanks

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into never-copy-http-buffer

original-brownbear committed:
CR: don't optimize NIO transport for now

original-brownbear commented:
Jenkins run elasticsearch-ci/1 (some Docker thing ...)

Tim-Brooks reviewed (approved):
LGTM

original-brownbear commented:
Thanks Tim!

@ywelsch are you good with this one as well? :)

ywelsch commented on modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpRequest.java:
```diff
@@ -41,23 +43,34 @@
+    private final AtomicBoolean released;
+    private final FullHttpRequest request;
+    private final boolean pooled;
+    private BytesReference content;
```
should we make the access to this field and the released field concurrency-safe?

original-brownbear replied:
It's just `final` now, the `null` on `release` was pointless in hindsight (see other comment).

ywelsch commented on modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpRequest.java:
```diff
@@ -105,9 +118,35 @@ public String uri() {
+
+    @Override
+    public HttpRequest releaseAndCopy() {
+        assert content != null;
```
`assert contentRef != null && released.get() == false;`?

ywelsch replied:
Also what if `release()` is called concurrently to this?

original-brownbear replied:
Yea you're right, the mechanics of this are weird ... looking at this again there's no win in `null`ing out the `content` here anyway since the `request` field keeps a reference to the bytes anyway for `Unpooled` buffers and for pooled ones we release the bytes just fine.

ywelsch commented on modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpRequest.java:
```diff
@@ -41,23 +43,34 @@
+    }
+
+    private Netty4HttpRequest(FullHttpRequest request, int sequence, AtomicBoolean released) {
+        this(request, new HttpHeadersMap(request.headers()), sequence, released, request.content() instanceof UnpooledHeapByteBuf == false,
```
I wonder how reliable this check is. In particular, how do we make sure this is not just a NOOP and we always return false? Can we have a test somewhere for this?

original-brownbear replied:
Just tried adding a test for this and it turned out to be a little trickier than expected. We currently only expose setting the unpooled allocator via a system property so this would require some additional code changes to make it testable.

I'm fine dropping this as well though, the unpooled allocator is pretty rare and it may not be worth the complexity optimizing for it here (since we're only optimizing the small request size case anyway). I just figured, since doing this check constrained to just `UnpooledHeapByteBuf`
was essentially a freebie and it worked in manual testing I'd go for it? :)

ywelsch replied:
@tbrooks8 thoughts on how to test this and/or whether to drop this?

Could we assert e.g. if `NettyAllocator.useUnpooled()` is true, that we then have an instance of `UnpooledHeapByteBuf` here?

original-brownbear replied:
> Could we assert e.g. if NettyAllocator.useUnpooled() is true, that we then have an instance of UnpooledHeapByteBuf here?

Unfortunately no, because larger messages will often show up as `CompositeByteBuf` (which we'd have to deep-inspect to make sure they don't contain any pooled buffers).  What a test for this would want to do, is to check that the unpooled allocator produces "some" instances of `UnpooledHeapByteBuf` here.
-> If this makes us uncomfortable I'd suggest just dropping it for now and revisiting optimizing the unpooled case later (if we care).

original-brownbear replied:
I ditched this check in 9ab9b4a now, realizing that it may not be safe in all cases anyway because `UnpooledHeapByteBuf` isn't `final` so if a child overrides the `release` method for whatever reason we'd be in trouble potentially.

I think we can do a better job here anyway up-stream by simply making use of the knowledge that we're running the unpooled allocator. I'd optimize this in a follow-up.

Can you take another look @ywelsch ? :)

ywelsch reviewed (commented):
I've left some comments around some details, overall looking good though

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into never-copy-http-buffer

original-brownbear committed:
CR: make content handling less weird

original-brownbear commented:
Thanks Yannick. I simplified the handling of `content` in 1812be4 and simplified the concurrency of this object through that. It was kind of pointless to optimize the unpooled case with the setting of `null` this way as explained above.

Not sure about a test for the unpooled buffer case, given the effort required for that. Maybe we could do so (add tricky tests) in a follow-up and use the logic from https://github.com/elastic/elasticsearch/pull/44881/ that allows for more wholistically optimizing the `Unpooled` case across all message sizes?

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into never-copy-http-buffer

original-brownbear committed:
remove unpooled hack

ywelsch reviewed (approved):
LGTM. Let's have @tbrooks8 ok this again as well.

original-brownbear commented:
Link https://github.com/elastic/elasticsearch/issues/49699 which shows OOM from the copies avoided here most likely

original-brownbear commented:
Ping @tbrooks8  There shouldn't be much left to do here here since the changes from the last time you looked at it are minor :) Thanks!

Tim-Brooks reviewed (approved):
I guess I missed this in the last review cycle, but I don't understand why we completely reverted this for NIO?

I thought you were going the direction of always copying instead of doing the cast optimization.

I think we should still using the existing buffer when `allowsUnsafeBuffers` return true. It's just that we should not attempt to do the non-copy when `allowsUnsafeBuffers` return false and we think the buffer is `unpooled` (since we do no reliably know that).

I guess we can bring NIO back in a follow-up so I'll approve this.

original-brownbear commented:
@tbrooks8 thanks! 

> I guess I missed this in the last review cycle, but I don't understand why we completely reverted this for NIO?

:man_facepalming: ... I completely misread your comment on the NIO situation back in the intial review ... will open the follow up tomorrow :)

original-brownbear commented:
NIO version in https://github.com/elastic/elasticsearch/pull/49819 :)
